src/prefect/new_futures.py

```diff
@@ -0,0 +1,179 @@
+        if self._final_state:
+            return self._final_state
+        client = get_client(sync_client=True)
+        try:
+            task_run = cast(TaskRun, client.read_task_run(task_run_id=self.task_run_id))
+        except ObjectNotFound:
+            # We'll be optimistic and assume this task will eventually start
+            # TODO: Consider using task run events to wait for the task to start
+            return Pending()
+        return task_run.state or Pending()
+
+    @abc.abstractmethod
+    def wait(self, timeout: Optional[float] = None) -> None:
+        ...
+        """
+        Wait for the task run to complete. 
+        
+        If the task run has already completed, this method will return immediately.
+
+        Args:
+            - timeout: The maximum number of seconds to wait for the task run to complete.
+              If the task run has not completed after the timeout has elapsed, this method will return.
+        """
+
+    @abc.abstractmethod
+    def result(
+        self,
+        timeout: Optional[float] = None,
+        raise_on_failure: bool = True,
+    ) -> Any:
+        ...
+        """
+        Get the result of the task run associated with this future.
+
+        If the task run has not completed, this method will wait for the task run to complete.
+
+        Args:
+            - timeout: The maximum number of seconds to wait for the task run to complete.
+            If the task run has not completed after the timeout has elapsed, this method will return.
+            - raise_on_failure: If `True`, an exception will be raised if the task run fails.
+
+        Returns:
+            The result of the task run.
+        """
+
+
+class PrefectConcurrentFuture(PrefectFuture):
+    """
+    A Prefect future that wraps a concurrent.futures.Future. This future is used
+    when the task run is submitted to a ThreadPoolExecutor.
+    """
+
+    def wait(self, timeout: Optional[float] = None) -> None:
+        try:
+            result = self._wrapped_future.result(timeout=timeout)
+        except TimeoutError:
+            return
+        if isinstance(result, State):
+            self._final_state = result
+
+    def result(
+        self,
+        timeout: Optional[float] = None,
+        raise_on_failure: bool = True,
+    ) -> Any:
+        if not self._final_state:
+            future_result = self._wrapped_future.result(timeout=timeout)
+            if isinstance(future_result, State):
+                self._final_state = future_result
+            else:
+                return future_result
+
+        _result = self._final_state.result(
+            raise_on_failure=raise_on_failure, fetch=True
+        )
+        # state.result is a `sync_compatible` function that may or may not return an awaitable
+        # depending on whether the parent frame is sync or not
+        if inspect.isawaitable(_result):
+            _result = run_sync(_result)
+        return _result
```

Comment:
Result retrieval is a sharp edge in the new engine. `@sync_compatible` will return a coroutine for `State.result()` when called from a sync context if there is an active event loop. We likely need a sync implementation for `State.result()`.